### ISSUE TITLE
[LibOS] 'getsockopt' fully implemented? 

### ISSUE THREAD

**Issue by martinschmatz** [oscarlab/graphene#548]
_Background:_ 
Page https://github.com/oscarlab/graphene/wiki/Implemented-System-Calls lists the system call 'getsockopt' under the title "System calls that are fully implemented".

_Observations:_ 
- File LibOS/glibc-2.19/bits/socket.h specifies the socket-level options for 'getsockopt' and 'setsockopt' (lines 304ff). Among other options, it enums SO_REUSEPORT = 0x0200,  /* Allow local address and port reuse.  */
- File LibOS/shim/src/shim_syscalls.c defines the shim syscall intercept for 'getsockopt' and re-routes it to 'shim_do_getsockopt' (line 338).
- 'shim_do_getsockopt' is implemented in file LibOS/shim/src/sys/shim_socket.c (lines 1725ff). However, the option SO_REUSEPORT  seems not to be handled in 'shim_do_getsockopt'. 

_Question:_ 
Is the 'getsockopt' system call indeed fully implemented, in particular for SO_REUSEPORT option? 

_Hint:_ 
gRPC requires this.

**Comment by donporter** [comment]
I'm not sure off the top of my head, but it sounds like this is a missing feature.

**Comment by dimakuv** [comment]
The `getsockopt()` syscall is *not* fully implemented: it doesn't cover *all* possible socket options. It probably makes sense to move `getsockopt/setsockopt` to the list of "System calls that are partially implemented"...

Concerning your specific option `SO_REUSEPORT`. This option is a kernel-level performance optimization for multi-threaded servers (see https://lwn.net/Articles/542629/). It doesn't make much sense to implement this option in Graphene-SGX because the host kernel is not trusted. Therefore, Graphene-SGX 
returns an error `ENOPROTOOPT`.

So no, `SO_REUSEPORT` is not implemented and there are no plans to implement it in near future (it just doesn't make sense in the SGX threat model).

To solve your problem with gRPC, you should try `grpc.so_reuseport` (e.g., `options=(('grpc.so_reuseport', 0),))` when creating a server.

**Comment by martinschmatz** [comment]
While it might not make sense to use the SO_REUSEPORT functionality in the kernel, I think this socket option should nevertheless be handled in a more meaningful way, e.g. instead of just returning -ENOPROTOOPT, an indication ('0', false,...) that SO_REUSEPORT is disabled could make more sense. 
Reason: With the current implementation of how SO_REUSEPORT is handled (= not handled), gRPC in the sync version of its c++ example 'helloworld' fails (even with builder.AddChannelArgument(GRPC_ARG_ALLOW_REUSEPORT, 0); - but thanks for this hint anyway!)

Btw: For gRPC in C++, we initially used a sync server implementation - BUT we had to modify the gRPC source in a way that gRPC doesn't to touch neither SO_REUSEPORT, SO_REUSEADDR nor TCP_NODELAY to make it execute. However, we then immediately did run into issue https://github.com/oscarlab/graphene/issues/517 because threads are not cleaned up after dynamically being executed. 

So, for gRPC in C++ in Graphene, we had to use the async server mode, where a _fixed_ number of threads poll the runtime queue in a while(true){} loop (= the threads never end and hence can life without being cleaned up). With this approach, still using the patched gRPC source for the time being, gRPC worked in a Graphene enclave in a C++ application. Needless to say, we'd prefer not to patch gRPC.

**Comment by dimakuv** [comment]
I am not sure what you mean by "an indication that SO_REUSEPORT is disabled could make more sense".

The issue with `SO_REUSEPORT` is that it cannot be silently ignored inside Graphene because the user app wants to create several sockets on the same port. This is an illegal operation in Graphene (or at least should be). It is unclear how to emulate the multiplexing of M sockets on a single port, given that we cannot rely on the host OS. If we would simply ignore the flag, nasty unexpected behaviors would result.

I am also curious why your app still fails with the disabled REUSEPORT option.

**Comment by martinschmatz** [comment]
As far as I understand the source (LibOS/shim/src/sys/shim_socket.c, lines 1833ff), Graphene handles the SO_REUSEADDR such that it always returns '1' for getsockopt. 
I was wondering whether or not such a pre-determined answer (I guess it would be '0'?) when adding the handling of SO_REUSEPORT in getsockopt would indicate to the user that SO_REUSEPORT is not active. This is not the same as 'silently ignoring'. 

At the end, it might well boil down to an allocation of pain: It simply looks like gRPC can not handle -ENOPROTOOPT as answer to a getsockopt check for SO_REUSEPORT . This behavior could be addressed by gRPC, or by Graphene. 

Which brings me to 'why your app still fails with the disabled REUSEPORT option': 
gRPC seemingly performs checks of what is available and what not. These checks are executed irrespective of whether or not the GRPC_ARG_ALLOW_REUSEPORT channel argument is set to 0 or 1. If set to '0', there's just one check less performed according to my measurements. 
Similar checks are done by gRPC for SO_REUSEADDR and TCP_NODELAY. This is why only after disabling all of these three checks in the gRPC source, our test app succeeded. 

In other words, the SO_REUSEPORT seems to be only the tip of the iceberg (at least for use on combination with gRPC): Current Graphene implementation for handling of SO_REUSEADDR and TCP_NODELAY sockopts also are not consumable by gRPC (at least not in its c++ version we're currently using; don't know about other languages - but given the checks are done in the core of gRPC, I believe it is not limited to c++).

**Comment by dimakuv** [comment]
https://github.com/grpc/grpc/blob/269493f9e85fde18d04a0851eae9886b3aad7c74/src/core/lib/iomgr/socket_utils_common_posix.cc#L162

Based on the gRPC code, you could try building gRPC library either with `-USO_REUSEPORT` or with `-DGPR_MANYLINUX1`.

Graphene-SGX seems to support `TCP_NODELAY` and `SO_REUSEADDR` (the latter in a kinda mock way). Why did you have problems with these flags?

Lastly, gRPC indeed doesn't check for `-ENOPROTOOPT`. Uncool. But gRPC authors at least enabled a couple macros to circumvent this (I guess they had reasons not to support this error code).

Based on the above discussion, I still don't think Graphene needs to support or silently ignore `SO_REUSEPORT`.

**Comment by mkow** [comment]
Regarding fully implementing getsockopt we already have #467. The original reporter haven't replied ad the gRPC discussion for quite some time, so the only issue left here is the bug in the docs, for which I'll submit a PR in a moment.

**Comment by mkow** [comment]
Hmm, actually we have it implemented fully, LibOS just returns that this option is not enabled (if I see correctly). This issue can be closed, I think.
